core/src/main/java/com/yahoo/ycsb/Client.java

```diff
@@ -35,250 +37,302 @@
+     */
+    @Override
+    public void run()
+    {
+        final long startTimeMs=System.currentTimeMillis();
+        final long startTimeNanos = System.nanoTime();
+        long deadline = startTimeNanos + _sleeptimeNs;
+        long startIntervalMs=startTimeMs;
+        long lastTotalOps=0;
+
+        boolean alldone;
+
+        do
+        {
+            long nowMs=System.currentTimeMillis();
+
+            lastTotalOps = computeStats(startTimeMs, startIntervalMs, nowMs, lastTotalOps);
+
+            alldone = waitForClientsUntil(deadline);
+
+            startIntervalMs=nowMs;
+            deadline+=_sleeptimeNs;
+        }
+        while (!alldone);
+
+        // Print the final stats.
+        computeStats(startTimeMs, startIntervalMs, System.currentTimeMillis(), lastTotalOps);
+    }
+
+    /**
+     * Computes and prints the stats.
+     *
+     * @param startTimeMs The start time of the test.
+     * @param startIntervalMs The start time of this interval.
+     * @param endIntervalMs The end time (now) for the interval.
+     * @param lastTotalOps The last total operations count.
+     *
+     * @return The current operation count.
+     */
+    private long computeStats(final long startTimeMs, long startIntervalMs, long endIntervalMs,
+            long lastTotalOps) {
+        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss:SSS");
+
+        long totalops=0;
+
+        // Calculate the total number of operations completed.
+        for (ClientThread t : _clients)
+        {
+            totalops+=t.getOpsDone();
+        }
+
+
+        long interval=endIntervalMs-startTimeMs;
+        double curthroughput=1000.0*(((double)(totalops-lastTotalOps))/((double)(endIntervalMs-startIntervalMs)));
+
+
+        DecimalFormat d = new DecimalFormat("#.##");
+        String label = _label + format.format(new Date());
+
+        StringBuilder msg = new StringBuilder(label).append(" ").append(interval/1000).append(" sec: ");
+        msg.append(totalops).append(" operations; ");
+
+        if (totalops != 0) {
+            msg.append(d.format(curthroughput)).append(" current ops/sec; ");
+        }
+
+        msg.append(Measurements.getMeasurements().getSummary());
+
+        System.err.println(msg);
+
+        if (_standardstatus) {
+            System.out.println(msg);
+        }
+        return totalops;
+    }
+
+    /**
+     * Waits for all of the client to finish or the deadline to expire.
+     *
+     * @param deadline The current deadline.
+     *
+     * @return True if all of the clients completed.
+     */
+    private boolean waitForClientsUntil(long deadline) {
+        boolean alldone=false;
+        long now=System.nanoTime();
+
+        while( !alldone && now < deadline ) {
+            try {
+                alldone = _completeLatch.await(deadline-now, TimeUnit.NANOSECONDS);
+            }
+            catch( InterruptedException ie) {
+                // Handled by while loop.
```

Comment:
If we get interrupted, we ought to return false instead of continuing to wait. (Presuming that will cause us to exit soon)

Should also set the thread's interrupted status.